### PR TITLE
Enhance benchmarks & bug fixes

### LINKED ISSUE #703: Collect progressEnd timings in benchmarks
The current suite of benchmark experiments has become obsolete after the change to use delayed data in #624 

To remediate this, @wz1000 and I propose that the benchmark runner is extended to collect two timings per experiment:

1. The experiment conclusion, which is generally a user visible response (e.g. completions) and involves delayed data
2. The progressEnd notification, which includes the time taken to refresh the actual data (e.g. recompute the completions)

So, keep the current experiments, but collect more metrics.

### PR DESCRIPTION
**Pull request by pepeiborra**
Enhance the benchmark suite to measure user time, delayed work time and allocations.
Fixes bugs I found in the process.

Fixes #703 and is also relevant for #242

### CODE REVIEW

File: bench/lib/Experiments.hs [hunk 1]
```diff
@@ -328,27 +350,35 @@ runBench runSess Bench {..} = handleAny (\e -> print e >> return badRun)
       changeDoc doc [hygienicEdit]
       waitForProgressDone
 
-
     liftIO $ output $ "Running " <> name <> " benchmark"
     (runSetup, userState) <- duration $ benchSetup doc
-    let loop 0 = return True
-        loop n = do
+    let loop userWaits delayedWork 0 = return $ Just (userWaits, delayedWork)
```
Comment (ndmitchell) [inline bench/lib/Experiments.hs:355]: I love a good space leak :) - from a cursor read, it looks like the first two variables will be space leaks, but can't be sure - best to just shove a ! in front of all three.

File: bench/lib/Experiments.hs [hunk 2]
```diff
@@ -401,11 +431,14 @@ setup = do
...
-  case find ("maximum residency" `isInfixOf`) (reverse $ lines input) of
-    Just l -> read $ filter isDigit $ head (words l)
-    Nothing -> -1
+parseMaxResidencyAndAllocations :: String -> (Int, Int)
+parseMaxResidencyAndAllocations input =
+    (f "maximum residency", f "bytes allocated in the heap")
+  where
+    inps = reverse $ lines input
+    f label = case find (label `isInfixOf`) inps of
+        Just l -> read $ filter isDigit $ head (words l)
```
Comment (ndmitchell) [inline bench/lib/Experiments.hs:440]: Fixing bracket and $ is a bit odd.

```suggestion
        Just l -> read $ filter isDigit $ head $ words l
```

Or you might want to use `word1` from Extra.

File: session-loader/Development/IDE/Session.hs [hunk 3]
```diff
@@ -245,11 +245,21 @@ loadSession dir = do
...
-
-    let consultCradle :: Maybe FilePath -> FilePath -> IO ([NormalizedFilePath], (IdeResult HscEnvEq, [FilePath]))
+          -- Typecheck all files in the project on startup
+          unless (null cs || not checkProject) $ do
+                cfps' <- liftIO $ filterM (IO.doesFileExist . fromNormalizedFilePath) (concatMap targetLocations cs)
+                void $ shakeEnqueue extras $ mkDelayedAction "InitialLoad" Debug $ void $ do
+                    mmt <- uses GetModificationTime cfps'
+                    let cs_exist = catMaybes (zipWith (<$) cfps' mmt)
+                    modIfaces <- uses GetModIface cs_exist
+                    -- update xports map
```
Comment (ndmitchell) [inline session-loader/Development/IDE/Session.hs:255]:
```suggestion
                    -- update exports map
```

**Review by ndmitchell — APPROVED** [review]

File: bench/lib/Experiments.hs (hunk 1 above)
Comment (pepeiborra) [reply bench/lib/Experiments.hs:355]: good catch!

Comment (pepeiborra) [reply bench/lib/Experiments.hs:355]: Although hopefully the GHC strictness analyser would figure that out as well. I wonder if that would make a good hover provider.

**Pull request merged**